pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -353,6 +395,58 @@ func logbackFromFile(t string, data interface{}) (string, error) {
 	return output.String(), nil
 }
 
+func SetStatusFailed(master *alpha1.CDAPMaster, message string) {
+	master.Status.UpgradeJobStartTimeNano = 0
+	master.Status.Meta.SetCondition(upgradeFailed, message, message)
+}
+
+// Adds a component list object in state in progress. This will cause UpdateStatus to update the
+// ready type to state "false".
+func AddUpgradeComponentNotReady(master *alpha1.CDAPMaster) {
+	os := status.ObjectStatus{}
+	os.Name = "upgrade"
+	os.Status = status.StatusInProgress
+	master.Status.ComponentMeta.ComponentList.Objects = append(master.Status.ComponentMeta.ComponentList.Objects, os)
+}
+
+func LaunchJob(master *alpha1.CDAPMaster, rsrclabels map[string]string) ([]reconciler.Object, error) {
+	master.Status.UpgradeJobStartTimeNano = time.Now().UnixNano()
+	master.Status.Meta.ClearCondition(upgradeFailed, upgradeStartMessage, upgradeStartMessage)
+
+	upgradeResources, err := GetJobResources(master, rsrclabels)
+	if err != nil {
+		SetStatusFailed(master, upgradeFailedInitMessage)
+		return nil, err
+	}
+	return upgradeResources, nil
+}
+
+// Gets the job resources and sets status to not ready, since any time when we wish to add the job
+// to expected, the status is not ready.
+func GetJobResources(master *alpha1.CDAPMaster, rsrclabels map[string]string) ([]reconciler.Object, error) {
```

Comment:
Seems this is quite specific for the pre-upgrade job. I think it's better to call it `getPreUpgradeJobResources`, which I also think should be a private function instead of public